include/mbgl/style/function/type.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <vector>
+#include <mbgl/util/optional.hpp>
+#include <mbgl/util/variant.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+namespace type {
+
+class Primitive;
+class Variant;
+class Array;
+class NArgs;
+class Typename;
+class Lambda;
+
+using ValueType = variant<
```

Comment:
This encoding of types looks odd to me.

* Why do we need both `ValueType` and `Type`?
* Is it necessary to divide types into "primitives" and other types? That seems like an unnecessary division.

Here's my sketch for how I'd do this. First, the types of values:

```C++
class Null {};
class Number {};
class String {};
class Color {};
class Object {};
class Array {
    Type itemType;
    optional<std::size_t> length;
};
class Value {};

using Type = variant<
    Null,
    Number,
    String,
    Color,
    Object,
    Array,
    Value>;
```

Then, you have special stuff for typing lambdas:

```C++
class Typename {
    std::string name;
};

class NArgs {
    std::vector<variant<Type, Typename>> types;
};

class Variant {
    std::vector<Type> choices;
};

using ResultType = variant<
    Type,
    Typename>;

using ArgumentType = variant<
    Type,
    NArgs,
    Variant>;

class LambdaExpression {
    ResultType resultType;
    std::vector<ArgumentType> argumentTypes;
}
```

`Expression::type` should be the type of the expression's result, not its lambda type. (When we talk about the "type of an expression" in any language, we're almost always referring to the type of its result value, not the type of the method if it's a method call.)

These suggestions apply to the JS implementation as well.